pkg/controller/apischeme/apischeme_controller.go

```diff
@@ -96,5 +100,300 @@ func (r *ReconcileAPIScheme) Reconcile(request reconcile.Request) (reconcile.Res
+		return reconcile.Result{}, err
+	}
+
+	clusterBaseDomain, err := utils.GetClusterBaseDomain(r.client)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't obtain the cluster's base domain")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't get the cluster's base domain", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+	masterNodeInstances, err := utils.GetClusterMasterInstances(r.client)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't detect the AWS instances for master nodes")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't find the cluster's AWS instances for master nodes", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+
+	// In theory this could return more than one VPC for master nodes. That would be
+	// odd since the security group is associated with a single VPC.
+	vpcs, err := utils.GetMasterNodeVPCs(r.client)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't get the VPC in use by master nodes")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't get the VPC id for master nodes", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+	if len(vpcs) > 1 {
+		reqLogger.Info(fmt.Sprintf("Multiple VPCs detected for master nodes. Using %s for security group", vpcs[0]))
+	}
+	// Now try to make sure all the things for Admin API are present in AWS
+	err = ensureAdminAPIEndpoint(reqLogger, instance, awsClient, config.AdminAPIName, config.AdminAPISecurityGroupName, vpcs[0], clusterBaseDomain, subnets, masterNodeInstances, ownerTags)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't ensure the admin API endpoint")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't ensure the admin API endpoint: "+err.Error(), cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+
+	// And now, tell the APIServer/cluster object about it.
+	err = r.addAdminAPIToAPIServerObject(reqLogger, config.AdminAPIName+"."+clusterBaseDomain)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't update APIServer/cluster object")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't update APIServer/cluster object", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+	SetAPISchemeStatus(reqLogger, instance, "Success", "Admin API Endpoint created", cloudingressv1alpha1.ConditionReady)
+	r.client.Status().Update(context.TODO(), instance)
+
+	return reconcile.Result{}, nil
+}
+
+func ensureCloudLoadBalancer(reqLogger logr.Logger, awsAPI *awsclient.AwsClient, endpointName string, elbSubnets []string, ownerTags map[string]string) (*awsclient.AWSLoadBalancer, error) {
```

Comment:
This function and ones below have a lot of shared arguments. Thoughts about wrapping them all in a struct?

Reply:
Yes, this will happen at some point, likely as part of an effort to support multiple cloud providers (they would each need a different such "data interface"). I'll have a look to see what effort is involved in creating a struct for all the functions/methods.